compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java

```diff
@@ -886,7 +885,11 @@ private void checkSupportedConfigType(BType type, Location location, String varN
         }
     }
 
-    private boolean isSupportedConfigType(BType type, List<String> errors, String varName) {
+    private boolean isSupportedConfigType(BType type, List<String> errors, String varName,
+                                          Set<BType> unresolvedTypes) {
+        if (!unresolvedTypes.add(type) && isRecordType(type)) {
```

Comment:
IMO we do not need a record type check. There can be cyclic union types as well.